twoliter/src/lock.rs

```diff
@@ -139,6 +249,124 @@ impl Lock {
+            .context("failed to deserialize manifest list")?;
+        let docker_arch = Architecture::try_from(arch)?;
+        manifest_list
+            .manifests
+            .iter()
+            .find(|x| x.platform.as_ref().unwrap().architecture == docker_arch)
+            .cloned()
+            .context(format!(
+                "could not find kit image for architecture '{}' at {}",
+                docker_arch, image.source
+            ))
+    }
+
+    async fn pull_image<P>(&self, path: P, digest_uri: &str) -> Result<IndexView>
+    where
+        P: AsRef<Path>,
+    {
+        let oci_archive = docker!(
+            ["save", digest_uri],
+            format!("failed to fetch kit and save to disk from {}", digest_uri)
+        );
+        let mut oci_archive = Archive::new(oci_archive.as_slice());
+        oci_archive
+            .unpack(path.as_ref())
+            .context("failed to unpack oci image")?;
+        let index_bytes = read(path.as_ref().join("index.json")).await?;
+        serde_json::from_slice(index_bytes.as_slice())
+            .context("failed to deserialize oci image index")
+    }
+
+    async fn extract_kit<P>(&self, path: P, image: &LockedImage, arch: &str) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let tmp_dir =
+            tempdir().context("failed to create a temporary directory for kit extraction")?;
+        let vendor = image.vendor.clone();
+        let name = image.name.clone();
+        let docker_arch = match arch {
+            "x86_64" => Ok("amd64"),
+            "aarch64" => Ok("arm64"),
+            _ => Err(anyhow!("unsupported architecture '{}'", arch)),
+        }?;
+        let target_path = path.as_ref().join(format!("{vendor}/{name}/{arch}"));
+        create_dir_all(&target_path).await?;
+
+        // First get the manifest for the specific requested architecture
+        let manifest = self.get_manifest(image, docker_arch).await?;
+
+        // Now pull the image and extract it to a temporary folder returning
+        // the oci index
+        let image_source = image.digest_uri(manifest.digest.as_str());
+        let index = self
+            .pull_image(tmp_dir.path(), image_source.as_str())
+            .await?;
+
+        // Read the manifest so we can get the layer digests
+        let digest = index
+            .manifests
+            .first()
+            .context("empty oci image")?
+            .digest
+            .replace(':', "/");
+        let manifest_bytes = read(tmp_dir.path().join(format!("blobs/{digest}")))
+            .await
+            .context("failed to read manifest blob")?;
+        let manifest_layout: ManifestLayoutView = serde_json::from_slice(manifest_bytes.as_slice())
+            .context("failed to deserialize oci manifest")?;
+
+        // Extract each layer into the target directory
+        for layer in manifest_layout.layers {
+            let digest = layer.digest.to_string().replace(':', "/");
```

Comment:
If we're just splitting the string on `algo:digest` - let's do a split instead.

Reply:
I did a replace cause this cleanly creates the path for use, i can do a split join if you would prefer?

Reply:
No, this is fine if there's a reason.